File: frontend/src/lib/api/comparison.ts
import { apiClient } from './client'
import { type FlowEdge, type FlowNode } from './flows'
import { type Scenario, type SimulationResult } from './simulations'

export type FlowSnapshot = {
  nodes: FlowNode[]
  edges: FlowEdge[]
}

export type ScenarioDiff = {
  scenario_id: string
  scenario_text: string
  action_before: string
  action_after: string
  changed: boolean
  trace_before_id: string | null
  trace_after_id: string | null
}

export type CompareResponse = {
  id: string
  project_id: string
  before_results: SimulationResult[]
  after_results: SimulationResult[]
  diff: ScenarioDiff[]
}

export type CompareRequest = {
  scenarios: Scenario[]
  flow_before: FlowSnapshot
  flow_after: FlowSnapshot
}

export function runComparison(
  projectId: string,
  request: CompareRequest,
): Promise<CompareResponse> {
  return apiClient.post<CompareResponse>(
    `/projects/${projectId}/simulations/compare`,
    request,
  )
}

File: backend/app/schemas/comparison.py
from pydantic import BaseModel
from app.schemas.flow import FlowUpdate
from app.schemas.simulation import Scenario, SimulationResult


class CompareRequest(BaseModel):
    scenarios: list[Scenario]
    flow_before: FlowUpdate
    flow_after: FlowUpdate


class ScenarioDiff(BaseModel):
    scenario_id: str
    scenario_text: str
    action_before: str
    action_after: str
    changed: bool
    trace_before_id: str | None = None
    trace_after_id: str | None = None


class CompareResponse(BaseModel):
    id: str
    project_id: str
    before_results: list[SimulationResult]
    after_results: list[SimulationResult]
    diff: list[ScenarioDiff]
